[ApiDiff] Implementation of the new ApiDiff that reuses GenAPI features

carlossanlop opened the pull request:
Follow up of https://github.com/dotnet/sdk/pull/46424

This PR is part of the work needed to create an ApiDiff tool that reuses some of the code from Microsoft.DotNet.GenAPI. The idea is to make the larger PR smaller and make it easier to review: https://github.com/dotnet/sdk/pull/45389

The proposed changes in this PR include:

- Full implementation of the tool that can generate an API diff from two assemblies.

carlossanlop committed:
ApiDiff implementation

carlossanlop committed:
Resources

carlossanlop committed:
Tests

carlossanlop committed:
Some infra files

ViktorHofer commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff.Tool/Microsoft.DotNet.ApiDiff.Tool.csproj:
```diff
@@ -0,0 +1,22 @@
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="System.CommandLine" VersionOverride="2.0.0-beta4.22272.1" />
```
Why the version override?

carlossanlop replied:
This is the latest version, it's the one documented in the official docs, and it has a completely different API surface (they introduce public API breaking changes).

We should consider updating our other Compat tools to using this version. I opened an issue to track this: https://github.com/dotnet/sdk/issues/46482

ViktorHofer replied:
Why not use the same one as the rest of the repo, apicompat and genapi? https://github.com/dotnet/sdk/blob/4df298b776b2f691b5b92e334104d584c722ef43/src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompat.Tool/Microsoft.DotNet.ApiCompat.Tool.csproj#L19 & https://github.com/dotnet/sdk/blob/4df298b776b2f691b5b92e334104d584c722ef43/src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI.Tool/Microsoft.DotNet.GenAPI.Tool.csproj#L13

carlossanlop replied:
It's too old. I think we should be testing latest whenever possible. I don't think it hurts to override the version. Eventually all other tools will have to use latest, ideally.

carlossanlop replied:
Updated to consume the latest updated last week. I had to also update the usage because the API surface changed considerably.

ViktorHofer commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff.Tool/Microsoft.DotNet.ApiDiff.Tool.csproj:
```diff
@@ -0,0 +1,22 @@
+  <PropertyGroup>
+    <TargetFramework>$(NetToolMinimum)</TargetFramework>
+    <OutputType>Exe</OutputType>
+    <!-- <IsPackable>true</IsPackable>
+    <PackAsTool>true</PackAsTool>
+    <ToolCommandName>genapidiff</ToolCommandName> -->
```
These should get enabled?

carlossanlop replied:
Yes. I had to disable them because otherwise it throws an error when I try to test it directly.

carlossanlop replied:
It's working now. Not sure what was happening but the fact that it's supposed to be a tool, prevented me from running the exe directly. No longer happening.

ViktorHofer commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff/Microsoft.DotNet.ApiDiff.csproj:
```diff
@@ -0,0 +1,17 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>$(NetToolMinimum)</TargetFrameworks>
```
```suggestion
    <TargetFramework>$(NetToolMinimum)</TargetFramework>
```

carlossanlop replied:
If only we had some sort of analyzer to catch these.

carlossanlop replied:
Fixed.

ViktorHofer commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff/Microsoft.DotNet.ApiDiff.csproj:
```diff
@@ -0,0 +1,17 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Verify.DiffPlex" />
```
Our product components usually don't depend on external libraries. Can you please explain why this dependency is necessary? Is that dependency available in source-build?

carlossanlop replied:
The ApiDiff tool does not need to be part of source-build.

The DiffPlex tool is heavily used in many other places in our dotnet repos. In fact, many folks from our team are often contributing to the repo of DiffPlex.

This tool is used to generate the markdown documents containing the diffs. So it's critical for the purpose of this tool.

ViktorHofer commented on test/Microsoft.DotNet.ApiDiff.Tests/Microsoft.DotNet.ApiDiff.Tests.csproj:
```diff
@@ -0,0 +1,25 @@
+
+  <PropertyGroup>
+    <TargetFrameworks>$(ToolsetTargetFramework)</TargetFrameworks>
+    <OutputType Condition="'$(TargetFramework)' == '$(ToolsetTargetFramework)'">Exe</OutputType>
```
```suggestion
    <OutputType>Exe</OutputType>
```

carlossanlop replied:
Fixed. But note that I copied this from somewhere else in this repo.

ViktorHofer commented on test/Microsoft.DotNet.ApiDiff.Tests/Microsoft.DotNet.ApiDiff.Tests.csproj:
```diff
@@ -0,0 +1,25 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>$(ToolsetTargetFramework)</TargetFrameworks>
```
```suggestion
    <TargetFramework>$(ToolsetTargetFramework)</TargetFramework>
```

carlossanlop replied:
Fixed.

ViktorHofer commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff.Tool/Microsoft.DotNet.ApiDiff.Tool.csproj:
```diff
@@ -0,0 +1,22 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\..\GenAPI\Microsoft.DotNet.GenAPI\Microsoft.DotNet.GenAPI.csproj" />
```
I was under the assumption that types were pushed down into ApiSymbolExtensions so that this new tool doesn't depend on GenAPI directly. Maybe I was wrong. Can you please tell how much code is used from that code base?

Would it make sense to create a new class library that is used by both GenAPI and ApiDiff so that this dependency is avoided?

ViktorHofer replied:
I think this isn't failing in source-build as in that build mode, the root sln doesn't get built. Instead a solution filter file gets used which doesn't have a reference to this new tool: https://github.com/dotnet/sdk/blob/d6bc7918c0be5005426fbbff43dfb622fc295142/eng/DotNetBuild.props#L14

Do we need to have this tool available when building source-only?

carlossanlop replied:
No, this tool isn't needed for source build. It's meant for manual use.

ViktorHofer replied:
Cool. Please also respond to my first comment in this discussion about the layering of class libraries.

carlossanlop replied:
> Can you please tell how much code is used from that code base?

It's basically the code that I had to refactor in my previous PRs. You can refer to those.

> Would it make sense to create a new class library that is used by both GenAPI and ApiDiff so that this dependency is avoided?

I'd like to understand this better. I'm not sure how this would be a problem, since ApiDiff won't be used in source-build, but takes a dependency on GenAPI.

ViktorHofer commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff.Tool/Program.cs:
```diff
@@ -0,0 +1,172 @@
+/// </summary>
+public static class Program
+{
+    private static readonly string[] defaultAttributesToExclude = new[]
+    {
+        "T:System.AttributeUsageAttribute",
+        "T:System.ComponentModel.EditorBrowsableAttribute",
+        "T:System.Diagnostics.CodeAnalysis.RequiresDynamicCodeAttribute",
+        "T:System.Diagnostics.CodeAnalysis.RequiresUnreferencedCodeAttribute"
+    };
```
IMO it feels wrong to have these defaults defined in the UI layer which should be as thin as possible. Could those be moved into the business layer and used if no attribute option is provided?

carlossanlop replied:
Done.

ViktorHofer commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff/DiffConfiguration.cs:
```diff
@@ -0,0 +1,21 @@
+/// <summary>
+/// Defines a variety of configuration options for API diff.
+/// </summary>
+public class DiffConfiguration
```
Looks a good example of a POCO. Should this be a record to be more lightweight?

carlossanlop replied:
Done.

ViktorHofer commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff.Tool/Program.cs:
```diff
@@ -0,0 +1,172 @@
+            diffConfig.AddPartialModifier,
+            diffConfig.HideImplicitDefaultConstructors);
+
+        Directory.CreateDirectory(diffConfig.OutputFolderPath);
+        foreach ((string assemblyName, string text) in results)
+        {
+            string filePath = Path.Combine(diffConfig.OutputFolderPath, $"{assemblyName}.md");
+            File.WriteAllText(filePath, text);
+            log.LogMessage($"Wrote '{filePath}'.");
+        }
```
Similar comment as above. I think it would make sense to move this into the business layer to keep the UI layer as thin as possible.

carlossanlop replied:
Mmm let me see what would be the best way to do this, as the intention was to be able to avoid writing to disk in the unit tests.

carlossanlop replied:
I did it by just adding another public method that runs the tool by writing to disk.

carlossanlop replied:
I'm exploring using a strategy as suggested.

carlossanlop replied:
Ok, I implemented it.

ViktorHofer commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff/DiffGenerator.cs:
```diff
@@ -0,0 +1,610 @@
+    private readonly bool _hideImplicitDefaultConstructors;
+    private readonly ISymbolFilter _symbolFilter;
+    private readonly ISymbolFilter _attributeSymbolFilter;
+    private readonly SyntaxTriviaList TwoSpacesTrivia;
+    private readonly SyntaxList<AttributeListSyntax> EmptyAttributeList;
```
These two fields should start with an `_` to match the other fields.

carlossanlop replied:
Done.

carlossanlop commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff.Tool/Program.cs:
```diff
@@ -0,0 +1,172 @@
+        }
+    }
+
+    private static void WaitForDebugger()
+    {
+        while (!Debugger.IsAttached)
+        {
+            Console.WriteLine($"Attach to process {Environment.ProcessId}...");
+            Thread.Sleep(1000);
+        }
+        Console.WriteLine("Debugger attached!");
+        Debugger.Break();
+    }
```
I do want to keep it. It only gets called when the user passed -D.

ViktorHofer replied:
Yes this should stay in the UI layer if you want to keep it.

carlossanlop commented on CODEOWNERS:
```diff
@@ -85,10 +85,15 @@
 
```
The errors in this file are unrelated to my changes.

carlossanlop committed:
Address some of the feedback

carlossanlop committed:
Move the IO code to the business layer. Add a test that verifies the method that reads/writes to disk works.

carlossanlop commented:
I just notice I didn't implement the addition of the table of contents file.

michaelgsharp reviewed (commented):
Gonna look over the main file some more but here are my initial thoughts.

carlossanlop commented on src/Compatibility/ApiDiff/Directory.Build.props:
```diff
@@ -0,0 +1,9 @@
+  <Import Project="..\Directory.Build.props" />
+
+  <PropertyGroup>
+    <SkipMicrosoftCodeAnalysisCSharpPinning>true</SkipMicrosoftCodeAnalysisCSharpPinning>
```
Without this property, this error shows up when building the apidif.slnf:

```
   C:\Users\calope\source\repos\sdk2\src\Compatibility\ApiDiff\Microsoft.DotNet.ApiDiff\Microsoft.DotNet.ApiDiff.csproj : error NU1608: Warning As Error: Detected package version outside of dependency constraint: Microsoft.CodeAnalysis.CSharp 4.4.0 requires Microsoft.CodeAnalysis.Common (= 4.4.0) but version Microsoft.CodeAnalysis.Common 4.14.0-1.25070.2 was resolved.
    C:\Users\calope\source\repos\sdk2\src\Compatibility\ApiDiff\Microsoft.DotNet.ApiDiff.Tool\Microsoft.DotNet.ApiDiff.Tool.csproj : error NU1608: Warning As Error: Detected package version outside of dependency constraint: Microsoft.CodeAnalysis.CSharp 4.4.0 requires Microsoft.CodeAnalysis.Common (= 4.4.0) but version Microsoft.CodeAnalysis.Common 4.14.0-1.25070.2 was resolved.
    C:\Users\calope\source\repos\sdk2\src\Compatibility\ApiDiff\Microsoft.DotNet.ApiDiff\Microsoft.DotNet.ApiDiff.csproj : error NU1109: 
      Detected package downgrade: Microsoft.CodeAnalysis.CSharp from 4.14.0-1.25070.2 to centrally defined 4.4.0. Update the centrally managed package version to a higher version. 
       Microsoft.DotNet.ApiDiff -> Microsoft.CodeAnalysis.CSharp.Workspaces 4.14.0-1.25070.2 -> Microsoft.CodeAnalysis.CSharp (= 4.14.0-1.25070.2)
       Microsoft.DotNet.ApiDiff -> Microsoft.CodeAnalysis.CSharp (>= 4.4.0)
    C:\Users\calope\source\repos\sdk2\src\Compatibility\ApiDiff\Microsoft.DotNet.ApiDiff.Tool\Microsoft.DotNet.ApiDiff.Tool.csproj : error NU1109:
      Detected package downgrade: Microsoft.CodeAnalysis.CSharp from 4.14.0-1.25070.2 to centrally defined 4.4.0. Update the centrally managed package version to a higher version.
       Microsoft.DotNet.ApiDiff.Tool -> Microsoft.CodeAnalysis.CSharp.Workspaces 4.14.0-1.25070.2 -> Microsoft.CodeAnalysis.CSharp (= 4.14.0-1.25070.2)
       Microsoft.DotNet.ApiDiff.Tool -> Microsoft.CodeAnalysis.CSharp (>= 4.4.0)
```

carlossanlop replied:
`Microsoft.CodeAnalysis.CSharp.Workspaces` is a brand new dependency I am introducing in this project for the first time in the whole repo.

carlossanlop commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff.Tool/Program.cs:
```diff
@@ -0,0 +1,159 @@
+            Description = "Hide implicit default constructors from types."
+        };
+
+        Option<bool> optionIncludeTableOfContents = new(["--includeTableOfContents", "-toc"], () => true)
```
Totally agree! I was working on that, since I also forgot to implement it: https://github.com/dotnet/sdk/pull/46425#issuecomment-2634834860

carlossanlop replied:
I decided to name it "--tableOfContentsTitle", as we _always_ want to generate a table of contents file.

carlossanlop commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff.Tool/Program.cs:
```diff
@@ -0,0 +1,159 @@
+            IsRequired = false
+        };
+
+        Option<bool> optionDebug = new(["--debug"], () => false)
```
Good point.

carlossanlop commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff.Tool/Program.cs:
```diff
@@ -0,0 +1,159 @@
+            Description = "Add the 'partial' modifier to types."
+        };
+
+        Option<string> optionAfterAssembliesFolderPath = new(["--after", "-a"])
```
old/new is probably a good option. The old diff tool uses old/new.

carlossanlop replied:
I thought a bit more about this, I think before vs after is fine. I think the words make sense for a differ.

carlossanlop commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff/DiffGenerator.cs:
```diff
@@ -0,0 +1,667 @@
+            case ChangeType.Modified:
+                Debug.Assert(beforeNode != null && afterNode != null);
+                string? changed = GenerateChangedDiff(beforeNodeWithoutAttributes!, afterNodeWithoutAttributes!);
+                // The attributes might have changed, but they were handled before, separately.
```
What the comment means here is that we already _printed_ as markdown whatever diffs were found for attributes. So at this stage, the never want to process attributes _again_.

carlossanlop replied:
Let me rewrite the comment, I think it isn't clear enough.

carlossanlop replied:
Fixed

carlossanlop commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff.Tool/Microsoft.DotNet.ApiDiff.Tool.csproj:
```diff
@@ -0,0 +1,22 @@
+    <OutputType>Exe</OutputType>
+    <IsPackable>true</IsPackable>
+    <PackAsTool>true</PackAsTool>
+    <ToolCommandName>genapidiff</ToolCommandName>
```
Ah, yes. Remnant from the initial name.

carlossanlop commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff/DiffGenerator.cs:
```diff
@@ -0,0 +1,667 @@
+                                                 bool hideImplicitDefaultConstructors,
+                                                IEnumerable<KeyValuePair<string, ReportDiagnostic>>? diagnosticOptions = null)
+    {
+        attributesToExclude = (attributesToExclude != null && attributesToExclude.Length > 0) ? attributesToExclude : DefaultAttributesToExclude;
```
good point

carlossanlop replied:
Fixed

carlossanlop committed:
Address most of the feedback

carlossanlop commented:
Addressed most of the feedback. Still need to go through a few more comments.

carlossanlop commented:
@MichaelSimons @ViktorHofer how do I exclude this src project from participating in SB?

I suspect that's the root cause of the failure saying that we are trying to sign the external dependency `Verify.DiffPlex` with a Microsoft certificate.

Note that `Verify.DiffPlex` is already used in many test projects in the sdk repo, which as we know, don't participate in SB. It seems this is the first src project that consumes `Verify.DiffPlex`.

MichaelSimons commented:
> @MichaelSimons @ViktorHofer how do I exclude this src project from participating in SB?
> 
> I suspect that's the root cause of the failure saying that we are trying to sign the external dependency `Verify.DiffPlex` with a Microsoft certificate.
> 
> Note that `Verify.DiffPlex` is already used in many test projects in the sdk repo, which as we know, don't participate in SB. It seems this is the first src project that consumes `Verify.DiffPlex`.
> 
> Do I just add `<DotNetBuildSourceOnly>false</DotNetBuildSourceOnly>` to the csprojs?

This appears to be affecting the linux unified build leg as well so excluding it from SB will not be enough.  From [sdk-unified-build (VMR Vertical Build Validation Ubuntu2404_DevVersions_x64)](https://github.com/dotnet/sdk/pull/46425/checks?check_run_id=36764983541)

```
    /__w/1/vmr/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Sign.proj(75,5): error SIGN004: Signing 3rd party library '/__w/1/vmr/src/sdk/artifacts/tmp/Release/ContainerSigning/339/tools/net8.0/any/Argon.dll' with Microsoft certificate 'MicrosoftDotNet500'. The library is considered 3rd party library due to its copyright: 'Copyright 2024. All rights reserved'.
    /__w/1/vmr/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Sign.proj(75,5): error SIGN004: Signing 3rd party library '/__w/1/vmr/src/sdk/artifacts/tmp/Release/ContainerSigning/340/tools/net8.0/any/DiffEngine.dll' with Microsoft certificate 'MicrosoftDotNet500'. The library is considered 3rd party library due to its copyright: 'Copyright 2024. All rights reserved'.
    /__w/1/vmr/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Sign.proj(75,5): error SIGN004: Signing 3rd party library '/__w/1/vmr/src/sdk/artifacts/tmp/Release/ContainerSigning/341/tools/net8.0/any/DiffPlex.dll' with Microsoft certificate 'MicrosoftDotNet500'. The library is considered 3rd party library due to its copyright: ''.
    /__w/1/vmr/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Sign.proj(75,5): error SIGN004: Signing 3rd party library '/__w/1/vmr/src/sdk/artifacts/tmp/Release/ContainerSigning/342/tools/net8.0/any/EmptyFiles.dll' with Microsoft certificate 'MicrosoftDotNet500'. The library is considered 3rd party library due to its copyright: 'Copyright 2024. All rights reserved'.
    /__w/1/vmr/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Sign.proj(75,5): error SIGN004: Signing 3rd party library '/__w/1/vmr/src/sdk/artifacts/tmp/Release/ContainerSigning/347/tools/net8.0/any/SimpleInfoName.dll' with Microsoft certificate 'MicrosoftDotNet500'. The library is considered 3rd party library due to its copyright: 'Copyright 2023. All rights reserved'.
    /__w/1/vmr/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Sign.proj(75,5): error SIGN004: Signing 3rd party library '/__w/1/vmr/src/sdk/artifacts/tmp/Release/ContainerSigning/358/tools/net8.0/any/Verify.dll' with Microsoft certificate 'MicrosoftDotNet500'. The library is considered 3rd party library due to its copyright: 'Copyright 2024. All rights reserved'.
    /__w/1/vmr/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Sign.proj(75,5): error SIGN004: Signing 3rd party library '/__w/1/vmr/src/sdk/artifacts/tmp/Release/ContainerSigning/359/tools/net8.0/any/Verify.DiffPlex.dll' with Microsoft certificate 'MicrosoftDotNet500'. The library is considered 3rd party library due to its copyright: 'Copyright 2024. All rights reserved'.
```

carlossanlop commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff.Tool/Program.cs:
```diff
@@ -0,0 +1,159 @@
+
+        // Custom ordering to match help menu.
+        log.LogMessage("Selected options:");
+            log.LogMessage($" - 'Before' assemblies:                {diffConfig.BeforeAssembliesFolderPath}");
```
Sure.
BTW that's a weird indentation.

carlossanlop commented:
> This appears to be affecting the linux unified build leg as well so excluding it from SB will not be enough. 

Unsure what to do then. I don't know how long it would take me to write our own diffing tool. Or maybe we could duplicate the diffplex source code if the license permits it.

ViktorHofer commented:
I think you just need to update Signing.props in this repo and tell the signing infrastructure how to sign the external assembly. This is by design afaik. @mmitche has more knowledge on this.

carlossanlop commented:
> I think you just need to update Signing.props in this repo and tell the signing infrastructure how to sign the external assembly. This is by design afaik. @mmitche has more knowledge on this.

Here!
https://github.com/dotnet/runtime/blob/16782a4481f389e68314b16ea628c5af80f22783/eng/Signing.props#L21-L22

carlossanlop committed:
Exclude diffplex DLLs from signing

carlossanlop commented:
@mmitche do I sign the packages like this? https://github.com/dotnet/sdk/pull/46425/commits/8a94049c91c5a0c41c1888e470b493174f237600

Or do I sign them using `$(ExternalCertificateId)` as indicated here? https://github.com/dotnet/sdk/blob/b513b97cdf0ecf1a3cbf35bd8900f18d75385a3f/eng/Signing.props#L55-L60

carlossanlop commented:
> Do I just add `<DotNetBuildSourceOnly>false</DotNetBuildSourceOnly>` to the csprojs?

@ViktorHofer @MichaelSimons assuming we can take care of the signing problem of DiffPlex, is this the right way to exclude a src project from source-build?

![image](https://github.com/user-attachments/assets/c802bef6-f581-4aa9-b74d-f6d33dba5e13)

MichaelSimons commented:
> > Do I just add `<DotNetBuildSourceOnly>false</DotNetBuildSourceOnly>` to the csprojs?
> 
> @ViktorHofer @MichaelSimons assuming we can take care of the signing problem of DiffPlex, is this the right way to exclude a src project from source-build?
> 
> ![image](https://private-user-images.githubusercontent.com/1175054/410594429-c802bef6-f581-4aa9-b74d-f6d33dba5e13.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg4NzMxODksIm5iZiI6MTczODg3Mjg4OSwicGF0aCI6Ii8xMTc1MDU0LzQxMDU5NDQyOS1jODAyYmVmNi1mNTgxLTRhYTktYjc0ZC1mNmQzM2RiYTVlMTMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDIwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAyMDZUMjAxNDQ5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDc3YWYwYjQ1YWViNjY0NGZmNzRjYjQ1ZTZhZDJjYWNhYWVjZjNlYWY0YWI4YjNmYmNkMzg2YThlZjU2YzE0MyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.NZKJlquZ0o6R40LTfDwKAcYbqUP3OUb55FK_2_2WMeA)

Yes - This is documented [here](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/Unified-Build-Controls.md#resource-controls)

carlossanlop commented:
Progress: setting DiffPlex resources to use no certificate got rid of the failure (pending confirming if it's the right choice, or if using `$(ExternalCertificateId)` is preferred, @mmitche).

The new failures are now related to the ApiDiff tool itself. Seems that the tests fail in Linux due to the wrong NewLine chars being used.

carlossanlop committed:
Fix markdown file title handling.
Disallow empty table of contents.
Add more file handling tests.
Add innerException when syntaxGenerator.Declaration does not work.

carlossanlop committed:
Exclude from source build.

carlossanlop committed:
Allow overriding attribute list, add tests.

carlossanlop committed:
improve attribute comment

carlossanlop commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff/DiffGenerator.cs:
```diff
@@ -0,0 +1,667 @@
+                                                 bool addPartialModifier,
+                                                 bool hideImplicitDefaultConstructors,
+                                                 IEnumerable<KeyValuePair<string, ReportDiagnostic>>? diagnosticOptions = null)
+    {
```
Fixed

carlossanlop committed:
Add tests for implicit and explicit operators.
Add a DISABLED test for explicit checked operator.

carlossanlop commented on test/Microsoft.DotNet.ApiDiff.Tests/Diff.Operators.Tests.cs:
```diff
@@ -0,0 +1,649 @@
+    }
+
+    [Fact]
+    public void TestExplicitOperator()
```
@ericstj this test passes in my local machine (should pass in the CI too, I just pushed it).

It's the next test (explicit checked) that fails.

ericstj replied:
That other one should be fixed now with `4.14.0-2.25111.1` of roslyn.  You could try updating AsmDiff to this version - it's safe to do so just for our tool projects / tests.  Don't update tasks, libs, or rest of SDK to that version though.

carlossanlop replied:
Fortunately, https://github.com/dotnet/sdk/pull/46546 was just merged so I don't need to update the packages manually.

carlossanlop replied:
Passing now!  Same with the next test, no longer skipped.

carlossanlop committed:
Implement API exclusion. Add tests.

carlossanlop commented on test/Microsoft.DotNet.ApiDiff.Tests/Diff.Operators.Tests.cs:
```diff
@@ -0,0 +1,760 @@
+    // we try to process it but we end up throwing an exception as it is unrecognized.
+    // It's going to be fixed with https://github.com/dotnet/roslyn/pull/77102
+    [Fact(Skip = "https://github.com/dotnet/roslyn/issues/77101")]
+    public void TestExcludeUnmodifiedOperator()
```
@ericstj I implemented API exclusion in the tool. Unfortunately, it is still failing because the API needs to be processed/recognized before determining if it needs to be excluded. I added a skipped test anyway.

carlossanlop replied:
Note that I added 3 other API exclusion tests above this one, but they are excluding non-checked operators, so those work fine.

ericstj replied:
> processed/recognized before determining if it needs to be excluded

I would hope we could feed the exclusion down into the filter used for genapi.  I think that should avoid trying to generate syntax for the API - since we can check if it's included before producing the syntax (just like we do for internal, for example).

carlossanlop replied:
Yes, let me look at at that. Looks like I missed something.

carlossanlop commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff/DiffGeneratorFactory.cs:
```diff
@@ -0,0 +1,112 @@
+    /// <summary>
+    /// The default attributes to exclude from the diff.
+    /// </summary>
+    public static readonly string[] DefaultAttributesToExclude = [
```
I think this can be improved by using the contents of this file by default: https://github.com/dotnet/runtime/blob/main/eng/DefaultGenApiDocIds.txt

Those are the attributes GenAPI excludes when generating the ref source files.

carlossanlop replied:
Maybe for now, this string array is enough, and I can improve this in a follow-up PR.

carlossanlop replied:
Created https://github.com/dotnet/sdk/issues/46922 to track it.

carlossanlop committed:
Remove Microsoft.CodeAnalysis.Worskpaces.Common PackageVersion from the compat Directory.Packages.props, it's not necessary because it's already in the global Versions.props and the global Versions.Details.xml.

carlossanlop committed:
Unskip the checked tests.

carlossanlop commented on src/Compatibility/Directory.Packages.props:
```diff
@@ -8,8 +8,4 @@
   </ItemGroup>
 
-  <ItemGroup>
-    <PackageVersion Include="Microsoft.CodeAnalysis.Workspaces.Common" Version="$(MicrosoftCodeAnalysisWorkspacesCommonPackageVersion)" />
```
@ericstj this was not needed anymore. It's in the Versions.props. I forgot to remove it in this PR, it was a temporary workaround while I merged the PR that introduced the new Versions.props and Version.Details.xml entries, which have been merged.

carlossanlop commented:
Looks like the remaining failures are just newline differences (`\r\n` vs `\n`). Working on that.

carlossanlop committed:
Add exception resource strings, make some methods static, ignore newline diffs in tests

carlossanlop committed:
fix the resources error

carlossanlop committed:
Make some code async. Add option to exclude assemblies. Add tests.

carlossanlop committed:
Suppress additional diagnostics by default

carlossanlop committed:
New resource

carlossanlop committed:
Fix bugs in attributes and events, add tests

carlossanlop committed:
remove static

carlossanlop committed:
Add event tests

carlossanlop committed:
Support attributes that take arguments, add tests.

carlossanlop committed:
Truly async

carlossanlop committed:
Fix bug with abstract events where the field event declaration needs to be obtained from the list of declaration variables.

carlossanlop committed:
Add test that confirms that no files are getting overwritten when multiple assemblies share a namespace. Fix concurrency issue.

carlossanlop committed:
Add arguments for before and after friendly name, separate that from the table of contents name.

carlossanlop committed:
Remove the body from methods and properties as requested in API diff PR.

carlossanlop committed:
Add unchanged assembly and unchanged namespace tests.

carlossanlop committed:
Add destructor tests.

carlossanlop committed:
Add field list tests.

carlossanlop committed:
Add an Options class for CSharpAssemblyDocumentGenerator to avoid growing the constructor to ridiculous sizes. Need this also for a cleaner selection of actions to take when generating the document.

Also, simplify the document generation by moving the single line rewriter to get executed alongside the rest of the rewriters. That way I don't have to regenerate the document and then regenerate the root node multiple times.

carlossanlop committed:
Moved some code from the generator to rewriters. Much faster.
Still faililng when expecting fully qualified names.

carlossanlop commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff/MemoryOutputDiffGenerator.cs:
```diff
@@ -0,0 +1,669 @@
+            AdditionalAnnotations = [Formatter.Annotation] // Formatter is needed to fix some spacing
+        };
+
+        CSharpAssemblyDocumentGenerator docGenerator = new(_log, options);
```
Here's how I create my `CSharpAssemblyDocumentGenerator` instance. Note the configuration options I chose.

carlossanlop commented on src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/CSharpAssemblyDocumentGenerator.cs:
```diff
@@ -25,75 +25,45 @@ namespace Microsoft.DotNet.GenAPI;
     /// <param name="assemblySymbol">The assembly symbol that represents the loaded assembly.</param>
     /// <returns>The source code document instance of the specified assembly symbol.</returns>
-    public Document GetDocumentForAssembly(IAssemblySymbol assemblySymbol)
+    public async Task<Document> GetDocumentForAssemblyAsync(IAssemblySymbol assemblySymbol)
```
Something inside this method is still reducing the fully qualified names and removing namespace prefixes like `System.` or `Microsoft.`. I already set the `ReducerAsync` optional and removed the `Simplifier.Annotation` instance from `AdditionalAnnotations` list. See https://github.com/dotnet/sdk/pull/46425/files#r1984035962

carlossanlop committed:
Split test files into smaller ones

carlossanlop committed:
Fix bug in TypeDeclarationCSharpSyntaxRewriter causing an extra space to stay after removing the base type global::System.Object.

carlossanlop committed:
cons string t in AttributeNameSuffixRemover

carlossanlop committed:
Fix bug in GlobalPrefixRemover preventing the removal of just global (it was also removing System)

carlossanlop committed:
Fix bug in primitive conversion rewriter preventing the correct visitation

carlossanlop committed:
Fix bugs in body removal rewriter in accesors, operators and records

carlossanlop committed:
Add code specific for records in memory generator. Move the most debugged method to the top for easier finding.

carlossanlop committed:
Revert the TypeDeclarationCSharpSyntaxRewriter changes, they were wrong and breaking the whole world.

carlossanlop committed:
Don't truncate the attribute name, leave the type name full.
Finally fixed the records.

carlossanlop committed:
missing comments

carlossanlop committed:
Missing docs

carlossanlop committed:
Missing docs

carlossanlop committed:
Move private member under public ones

carlossanlop committed:
Replace the WaitAll with actual async method calls.

carlossanlop committed:
Better split of code

carlossanlop committed:
Fix sdk.sln merge conflict

carlossanlop commented:
@ViktorHofer @ericstj I pushed a ton of additional commits fixing several bugs I've been finding. I added a ton of tests too.

I think the tool is in a good shape for merging now. Anything else that we would want to address, I'd like to address it in separate PRs, if you don't mind.

So with that in mind, here is the ApiDiff generated with the tip of this PR: https://github.com/dotnet/core/pull/9821

ericstj commented on eng/Signing.props:
```diff
@@ -23,10 +23,19 @@
     <FileSignInfo Include="Castle.Core.dll" CertificateName="None" />
     <FileSignInfo Include="Moq.dll" CertificateName="None" />
 
+    <!-- Third party DLLs used in src projects not participating in source-build -->
```
Source-build isn't a differentiator here for signing, maybe it's OK to not sign since this is a non-shipping tool?  If it's shipping, then we should sign with the 3rd party cert `$(ExternalCertificateId)` as is done for other 3rd party bits below.

carlossanlop replied:
This dependency is used in product code, unlike the ones using `None` which are dependencies for test code. So it makes sense to use `$(ExternalCertificateId)` as you suggest.

ericstj commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff.Tool/Program.cs:
```diff
@@ -0,0 +1,209 @@
+
+        Option<string> optionBeforeRefAssembliesFolderPath = new(["--refbefore", "-rb"])
+        {
+            Description = "The path to the folder containing the old (before) reference assemblies.",
```
```suggestion
            Description = "The path to the folder containing the references required by old (before) assemblies, not to be included in the diff.",
```

carlossanlop replied:
Fixed

ericstj commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff.Tool/Program.cs:
```diff
@@ -0,0 +1,209 @@
+
+        Option<string> optionBeforeAssembliesFolderPath = new(["--before", "-b"])
+        {
+            Description = "The path to the folder containing the old (before) assemblies.",
```
```suggestion
            Description = "The path to the folder containing the old (before) assemblies to be included in the diff.",
```

carlossanlop replied:
Fixed

ericstj commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff.Tool/Program.cs:
```diff
@@ -0,0 +1,209 @@
+
+        Option<string> optionAfterAssembliesFolderPath = new(["--after", "-a"])
+        {
+            Description = "The path to the folder containing the new (after) assemblies.",
```
```suggestion
            Description = "The path to the folder containing the new (after) assemblies to be included in the diff.",
```

carlossanlop replied:
Fixed

ericstj commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff.Tool/Program.cs:
```diff
@@ -0,0 +1,209 @@
+
+        Option<string> optionAfterRefAssembliesFolderPath = new(["--refafter", "-ra"])
+        {
+            Description = "The path to the folder containing the new (after) reference assemblies.",
```
```suggestion
            Description = "The path to the folder containing references required by the new (after) reference assemblies, not to be included in the diff.",
```

carlossanlop replied:
Fixed

ericstj commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff.Tool/Program.cs:
```diff
@@ -0,0 +1,209 @@
+        {
+            Description = "The path to the folder containing the new (after) assemblies.",
+            Arity = ArgumentArity.ExactlyOne,
+            IsRequired = false
```
I would expect this to be required.
```suggestion
            IsRequired = true
```

carlossanlop replied:
Fixed

ericstj commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff.Tool/Program.cs:
```diff
@@ -0,0 +1,209 @@
+    {
+        RootCommand rootCommand = new("genapidiff");
+
+        Option<string> optionBeforeAssembliesFolderPath = new(["--before", "-b"])
```
General thought - do the inputs match our other compatibility tools?  Do they accept the same sort of things in the same format as other tools?

Suppose I might want to automate a diff in a library project I build.  Would I have to do unnatural things like copy my references to a single folder?

carlossanlop replied:
GenAPI/ApiCompat take as argument an array of folders. This tool only takes one folder.

The AssemblySymbolLoader `AddReferenceSearchPaths` method adds this single folder, and later the `LoadFromPaths` method enumerates the files, but it is not using `SearchOption.AllDirectories` so it won't recurse.

I see what you mean, I think we can easily change the behavior of the tool to take multiple folders, or visit recursively, if we find the need to do so. At the moment I haven't had the need to do that because the ref nupkgs contain all the DLLs in one single folder.

ericstj replied:
I was imagining someone might want to introduce this tool into their project build.  For example WinForms, EF, Aspire, etc.  Imagine that after running APICompat they also get a nice diff for release notes.  IMO it's easier for us to be consistent.  Being arbitrarily different creates friction.  Consider as a follow up.

carlossanlop replied:
Yeah you're right. I fixed it. It matches the other tools now.

ericstj commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff.Tool/Program.cs:
```diff
@@ -0,0 +1,209 @@
+        {
+            Description = "The title of the markdown table of contents file that is placed in the output folder.",
+            Arity = ArgumentArity.ExactlyOne,
+            IsRequired = true
```
Seems we could have a default for a lot of these things?  For instance if a single assembly were passed in we could probably come up with reasonable defaults.  Might be interesting to improve a lighter-weight "happy path"

carlossanlop replied:
Sure, I think I could use a simple default and make this argument optional. The powershell script we use for the monthly api diffs will continue passing a value relevant to the diff comparison being done.

ericstj replied:
Parameter is still marked as required.

carlossanlop replied:
Fixed

ericstj commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff.Tool/Program.cs:
```diff
@@ -0,0 +1,209 @@
+
+        Option<string[]?> optionAssembliesToExclude = new(["--assembliesToExclude", "-eas"], () => null)
+        {
+            Description = "Assemblies to exclude from the diff.",
```
```suggestion
            Description = "Assemblies from both before and after to exclude from the diff.",
```

carlossanlop replied:
Fixed

ericstj commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff.Tool/Program.cs:
```diff
@@ -0,0 +1,209 @@
+
+        Option<string[]?> optionAttributesToExclude = new(["--attributesToExclude", "-eattrs"], () => null)
+        {
+            Description = "Attributes to exclude from the diff.",
```
What's the format?  Is it a file with docIDs?  If not - can we put docIDs on the commandline safely?  They have things like `<` and `>` etc that might get sketchy.

carlossanlop replied:
I'll change this to instead receive a file path. The contents will then still be passed as an array of strings so this can easily be unit tested without involving files.

I'll also mention the format.

ericstj commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff.Tool/Program.cs:
```diff
@@ -0,0 +1,209 @@
+
+        Option<string[]?> optionApisToExclude = new(["--apisToExclude", "-eapis"], () => null)
+        {
+            Description = "APIs to exclude from the diff.",
```
Same question as above - what's the format?

carlossanlop replied:
Will do the same for this.

ericstj commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff.Tool/Program.cs:
```diff
@@ -0,0 +1,209 @@
+
+        Option<bool> optionHideImplicitDefaultConstructors = new(["--hideImplicitDefaultConstructors", "-hidc"], () => false)
+        {
+            Description = "Hide implicit default constructors from types."
```
Should this be an option?  I would expect we'd always want to show when a change removed a default constructor, or added one.  When would someone not want to see this?

carlossanlop replied:
I remember api diffs where area owners were wondering why a default constructor was being shown when no one had added one. But we still decided that we needed to keep showing them.

I suppose it's similar to api-docs, where we always show a default constructor even if one isn't explicitly defined in source code. I'll remove this as an option, since I agree with what you said that we probably will never disable them.

ericstj commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff/DiffGeneratorFactory.cs:
```diff
@@ -0,0 +1,125 @@
+        "T:System.AttributeUsageAttribute",
+        "T:System.ComponentModel.EditorBrowsableAttribute",
+        "T:System.Diagnostics.CodeAnalysis.RequiresDynamicCodeAttribute",
+        "T:System.Diagnostics.CodeAnalysis.RequiresUnreferencedCodeAttribute"
```
Do we really want to bake all these exclusions in?  They all seem meaningful human-authored attributes.

carlossanlop replied:
These are the ones I had to exclude in the old AsmDiff already. I was asked by area owners to remove them as they weren't important for the ApiDiff and they're quite noisy.

carlossanlop replied:
You're right. Removed from source code. I instead put them in a file that gets copied alongside the tool exe and I documented it in the argument description.

ericstj reviewed (commented):
Some initial feedback, mostly on UI.

carlossanlop committed:
Add enum support. Fix minor bug with records with members.

carlossanlop committed:
Fix bug preventing attributes with arguments from getting either suppressed or all kept (without a docId collision).

carlossanlop committed:
Shorter test names.

carlossanlop committed:
Yet one more attribute bug fix and test.

Rob-Hague commented:
Hi @carlossanlop thanks for working on this! I gave it a go on the SSH.NET repo and it worked very nicely. I have a couple of questions/comments on the output:

1. Adding a new interface shows up but does not display any members, nor inherited interfaces nor type constraints. Not sure if intended? Actual:
```diff
+     public interface IOrderedDictionary<TKey, TValue>
+     {
+     }
```
Expected something like:
```diff
+     public interface IOrderedDictionary<TKey, TValue> :
+        IDictionary<TKey, TValue>, IReadOnlyDictionary<TKey, TValue>
+        where TKey : notnull
+     {
+        bool ContainsValue(TValue value);
+     }
```
2. When a base constructor is changed then it shows in the diff of derived class constructors. The base constructor change was an API break so unlikely to happen in the BCL, but also not sure if this output (along with the `default()!` syntax) is intended
```diff
      public sealed class AesCipher : Renci.SshNet.Security.Cryptography.BlockCipher
      {
-         public AesCipher(byte[] key, byte[] iv, Renci.SshNet.Security.Cryptography.Ciphers.AesCipherMode mode, bool pkcs7Padding = false) : base(default(byte[])!, default(byte)!, default(Renci.SshNet.Security.Cryptography.Ciphers.CipherMode)!, default(Renci.SshNet.Security.Cryptography.Ciphers.CipherPadding)!);
+         public AesCipher(byte[] key, byte[] iv, Renci.SshNet.Security.Cryptography.Ciphers.AesCipherMode mode, bool pkcs7Padding = false) : base(default(byte[])!, default(byte)!, default(Renci.SshNet.Security.Cryptography.Ciphers.CipherMode)!, default(Org.BouncyCastle.Crypto.Paddings.IBlockCipherPadding)!);
      }
```

carlossanlop commented:
@Rob-Hague thanks for testing! It could be a bug but I can't tell for sure, I'm on my phone.

I think I have interface unit tests. Did I cover your scenario? If not, I can take a look on Monday.

Rob-Hague commented:
I see tests where an interface is added, but I did not find any where the interface has any members. Here is a change that I was expecting to pass, but fails:

```diff
diff --git a/test/Microsoft.DotNet.ApiDiff.Tests/Diff.Interface.Tests.cs b/test/Microsoft.DotNet.ApiDiff.Tests/Diff.Interface.Tests.cs
index a5c820e17c..d4b4a4be94 100644
--- a/test/Microsoft.DotNet.ApiDiff.Tests/Diff.Interface.Tests.cs
+++ b/test/Microsoft.DotNet.ApiDiff.Tests/Diff.Interface.Tests.cs
@@ -25,6 +25,8 @@ public struct MyStruct
                     }
                     public interface IMyInterface
                     {
+                        int MyMethod(string theParameter);
+                        long MyProperty { get; }
                     }
                 }
                 """,
@@ -33,6 +35,8 @@ namespace MyNamespace
                   {
                 +     public interface IMyInterface
                 +     {
+                +         int MyMethod(string theParameter);
+                +         long MyProperty { get; }
                 +     }
                   }
                 """);
```

but maybe it just a case of what is expected here.

As for my second point (the base class constructor) I think that is such an edge case that it is probably not worth looking at

carlossanlop committed:
Use ExternalCertificateId to sign DiffPlex dependencies.

carlossanlop committed:
Address UI feedback.

carlossanlop committed:
Take lists of files with elements to exclude, instead of directly passing list of elements to exclude.

carlossanlop committed:
Add missing case for including interface members.
Add tests to verify interface members get included.
Add test for interface type constraints, but needs more investigation.

carlossanlop commented:
@Rob-Hague I fixed the bug preventing interface members from getting added.
  
The interface type constraints does seem to work but it's adding more than expected: each of the inherited interfaces is 'exploded' into all the interfaces each one of them inherits. This is probably not an issue but I think the preference would be to only include what the user intended to include explicitly. Also, the nullability constraint is not showing up, so for that reason I marked the test as skipped. I'll have to investigate later (preferrably after I merge this PR).

I can also investigate further the problem you reported with the constructor, but does not seem to be a common case. So if you don't mind, I'd also like to investigate it later.

@ericstj I addressed all your feedback.

Forgind commented:
> Second [time](https://github.com/dotnet/sdk/pull/48034#issuecomment-2784595639) I see a bunch of duplicate license/cla legs here in the sdk repo. This time I got a screenshot @Forgind: ![image](https://private-user-images.githubusercontent.com/1175054/431493090-d386c701-9053-483f-b468-6e7a9e4ac78c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQxNTA0MjksIm5iZiI6MTc0NDE1MDEyOSwicGF0aCI6Ii8xMTc1MDU0LzQzMTQ5MzA5MC1kMzg2YzcwMS05MDUzLTQ4M2YtYjQ2OC02ZTdhOWU0YWM3OGMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MDhUMjIwODQ5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWFhZjVjYzVjYzc4OTJhMDdjMThlNDdjZDYyYjE3NDA4NmNiNWI0YTg0MjJmYjIzNzE2YWUwZmQ0NTNiNzAxMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.-q_7ObntTqbnai6cAV0ABZpRZUE9pJcRfk5VC3_qwVY)

Very strange. It looks like it disappeared again. I'm split between "it resolves itself, so it isn't a big deal" and "definitely a bug; we should ping first responders"

ericstj commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff.Tool/Microsoft.DotNet.ApiDiff.Tool.csproj:
```diff
@@ -0,0 +1,23 @@
+    <TargetFramework>$(NetToolMinimum)</TargetFramework>
+    <OutputType>Exe</OutputType>
+    <IsPackable>true</IsPackable>
+    <PackAsTool>true</PackAsTool>
```
Consider adding explicit IsShipping properties if we don't wish to ship this.

carlossanlop replied:
Done

ericstj commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff/xlf/Resources.cs.xlf:
```diff
@@ -0,0 +1,52 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<xliff xmlns="urn:oasis:names:tc:xliff:document:1.2" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" version="1.2" xsi:schemaLocation="urn:oasis:names:tc:xliff:document:1.2 xliff-core-1.2-transitional.xsd">
```
Since this is a non-shipping tool it might be better to disable localization rather than use resources to localize this.  Your call but I can't imagine a case for us to use this localization right now.

carlossanlop replied:
Makes sense. Removed.

ericstj reviewed (approved):
Please consider feedback and address or file follow-up. This can merge when you're happy with it as it is non-shipping tool at the moment.

ViktorHofer commented on src/Compatibility/ApiDiff/Microsoft.DotNet.ApiDiff.Tool/Properties/launchSettings.json:
Please delete this file before merging

carlossanlop replied:
Removed

carlossanlop committed:
Merge remote-tracking branch 'dotnet/main' into ApiDiff

carlossanlop committed:
Solve merge confict in sdk.sln

carlossanlop committed:
System.CommandLine override no longer needed.

carlossanlop committed:
Delete launchSettings.json

carlossanlop committed:
Update System.CommandLine usage to match latest version's API surface.

carlossanlop committed:
test typo

carlossanlop committed:
No resources needed yet

carlossanlop committed:
Remove Required from optioNTableOfContentsTitle

carlossanlop committed:
Set IsShippingPackage to false

carlossanlop committed:
The default attributes to exclude list should only apply to the ApiDiff Tool, not to the library source code. Modify the existing test to consume the same list from the tool.

carlossanlop committed:
Delete xlfs

carlossanlop committed:
Merge branch 'main' into ApiDiff
